File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.rs
impl Solution {
    pub fn max_profit(prices: Vec<i32>) -> i32 {
        let n = prices.len();
        let mut max_prices = vec![0; n];
        max_prices[n-1] = prices[n-1];
        for i in (0..n-1).rev() {
            max_prices[i] = max_prices[i+1].max(prices[i]); 
        }
        let mut max_profit = 0;
        for i in 0..n {
            max_profit = max_profit.max(max_prices[i] - prices[i]); 
        }
        max_profit
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.ts
function maxProfit(prices: number[]): number {
    let n = prices.length;
    let minPrice = prices[0];
    let maxProfit = 0;
    for (let i = 1; i < n; i++) {
        if (prices[i] < minPrice) {
            minPrice = prices[i];
        } else if (prices[i] - minPrice > maxProfit) {
            maxProfit = prices[i] - minPrice;
        }
    }

    return maxProfit;
};